File: native/platform/darwin/shortcuts.m
// Global shortcuts via Carbon RegisterEventHotKey. We pay the small
// cost of a Carbon dependency to get the right semantics — these
// hotkeys fire system-wide, no accessibility prompt, work even when
// the app is hidden. NSEvent.addGlobalMonitorForEvents would also
// work but only delivers events while the app is unfocused, which
// breaks the typical "summon window" hotkey use case.

#import <Cocoa/Cocoa.h>
#import <Carbon/Carbon.h>
#import "shortcuts.h"

// Forward declaration — defined in app/app_events.zc, dispatches an
// app event to native callbacks + webviews + workers.
extern int zapp_app_dispatch(int event_id, const char* data);

// We use a regular string event name rather than a numbered AppEvent —
// shortcuts aren't a system-defined lifecycle event, and the runtime
// subscribes via Events.on("app:shortcut-triggered", ...) anyway. This
// avoids a synchronized constant across events.zc / events.ts / bootstrap.
extern void darwin_webview_eval_all(const char* js);
extern void worker_broadcast_eval_js(char* js);

// --- Accelerator → Carbon constants ---

static UInt32 carbon_modifiers_from_token(NSString* token) {
    NSString* lower = [token lowercaseString];
    if ([lower isEqualToString:@"cmd"] || [lower isEqualToString:@"command"] ||
        [lower isEqualToString:@"cmdorctrl"] || [lower isEqualToString:@"meta"]) {
        return cmdKey;
    }
    if ([lower isEqualToString:@"ctrl"] || [lower isEqualToString:@"control"]) return controlKey;
    if ([lower isEqualToString:@"alt"]  || [lower isEqualToString:@"option"])  return optionKey;
    if ([lower isEqualToString:@"shift"]) return shiftKey;
    return 0;
}

// Map a key token (the part after the last "+") to a Carbon virtual
// key code. Returns UINT32_MAX on unknown.
static UInt32 carbon_keycode_for_token(NSString* token) {
    NSString* k = [token lowercaseString];
    if (k.length == 0) return UINT32_MAX;

    // Letters
    if (k.length == 1) {
        unichar c = [k characterAtIndex:0];
        switch (c) {
            case 'a': return kVK_ANSI_A; case 'b': return kVK_ANSI_B;
            case 'c': return kVK_ANSI_C; case 'd': return kVK_ANSI_D;
            case 'e': return kVK_ANSI_E; case 'f': return kVK_ANSI_F;
            case 'g': return kVK_ANSI_G; case 'h': return kVK_ANSI_H;
            case 'i': return kVK_ANSI_I; case 'j': return kVK_ANSI_J;
            case 'k': return kVK_ANSI_K; case 'l': return kVK_ANSI_L;
            case 'm': return kVK_ANSI_M; case 'n': return kVK_ANSI_N;
            case 'o': return kVK_ANSI_O; case 'p': return kVK_ANSI_P;
            case 'q': return kVK_ANSI_Q; case 'r': return kVK_ANSI_R;
            case 's': return kVK_ANSI_S; case 't': return kVK_ANSI_T;
            case 'u': return kVK_ANSI_U; case 'v': return kVK_ANSI_V;
            case 'w': return kVK_ANSI_W; case 'x': return kVK_ANSI_X;
            case 'y': return kVK_ANSI_Y; case 'z': return kVK_ANSI_Z;
            case '0': return kVK_ANSI_0; case '1': return kVK_ANSI_1;
            case '2': return kVK_ANSI_2; case '3': return kVK_ANSI_3;
            case '4': return kVK_ANSI_4; case '5': return kVK_ANSI_5;
            case '6': return kVK_ANSI_6; case '7': return kVK_ANSI_7;
            case '8': return kVK_ANSI_8; case '9': return kVK_ANSI_9;
        }
    }

    // Named keys
    if ([k isEqualToString:@"space"])       return kVK_Space;
    if ([k isEqualToString:@"tab"])         return kVK_Tab;
    if ([k isEqualToString:@"return"]
        || [k isEqualToString:@"enter"])    return kVK_Return;
    if ([k isEqualToString:@"escape"]
        || [k isEqualToString:@"esc"])      return kVK_Escape;
    if ([k isEqualToString:@"delete"]
        || [k isEqualToString:@"backspace"]) return kVK_Delete;
    if ([k isEqualToString:@"forwarddelete"]) return kVK_ForwardDelete;
    if ([k isEqualToString:@"left"])        return kVK_LeftArrow;
    if ([k isEqualToString:@"right"])       return kVK_RightArrow;
    if ([k isEqualToString:@"up"])          return kVK_UpArrow;
    if ([k isEqualToString:@"down"])        return kVK_DownArrow;
    if ([k isEqualToString:@"home"])        return kVK_Home;
    if ([k isEqualToString:@"end"])         return kVK_End;
    if ([k isEqualToString:@"pageup"])      return kVK_PageUp;
    if ([k isEqualToString:@"pagedown"])    return kVK_PageDown;
    if ([k isEqualToString:@"f1"])  return kVK_F1;  if ([k isEqualToString:@"f2"])  return kVK_F2;
    if ([k isEqualToString:@"f3"])  return kVK_F3;  if ([k isEqualToString:@"f4"])  return kVK_F4;
    if ([k isEqualToString:@"f5"])  return kVK_F5;  if ([k isEqualToString:@"f6"])  return kVK_F6;
    if ([k isEqualToString:@"f7"])  return kVK_F7;  if ([k isEqualToString:@"f8"])  return kVK_F8;
    if ([k isEqualToString:@"f9"])  return kVK_F9;  if ([k isEqualToString:@"f10"]) return kVK_F10;
    if ([k isEqualToString:@"f11"]) return kVK_F11; if ([k isEqualToString:@"f12"]) return kVK_F12;

    return UINT32_MAX;
}

// Parse "CmdOrCtrl+Shift+Space" → (modifiers, keycode). Returns false
// on any unknown component.
static bool parse_accelerator(const char* accelerator, UInt32* out_mods, UInt32* out_key) {
    if (!accelerator || !accelerator[0]) return false;
    NSString* s = [NSString stringWithUTF8String:accelerator];
    if (!s) return false;
    NSArray<NSString*>* parts = [s componentsSeparatedByString:@"+"];
    if (parts.count == 0) return false;

    UInt32 mods = 0;
    UInt32 key = UINT32_MAX;
    for (NSUInteger i = 0; i < parts.count; i++) {
        NSString* tok = [parts[i] stringByTrimmingCharactersInSet:
                            [NSCharacterSet whitespaceCharacterSet]];
        if (i == parts.count - 1) {
            // Last token is the key.
            key = carbon_keycode_for_token(tok);
        } else {
            UInt32 m = carbon_modifiers_from_token(tok);
            if (m == 0) return false;  // unknown modifier
            mods |= m;
        }
    }
    if (key == UINT32_MAX) return false;
    *out_mods = mods;
    *out_key = key;
    return true;
}

// --- Registration table ---
//
// Map accelerator string ↔ EventHotKeyRef. The Carbon ID we stuff into
// EventHotKeyID is just a sequential counter; the dispatch table in
// `slots[]` resolves the ID back to the accelerator string for the
// JS-side dispatch.

#ifndef ZAPP_MAX_SHORTCUTS
#define ZAPP_MAX_SHORTCUTS 64
#endif

typedef struct {
    BOOL active;
    UInt32 id;                  // Carbon EventHotKeyID
    EventHotKeyRef ref;         // Carbon registration handle
    char accelerator[128];      // owner-supplied string
} ShortcutSlot;

static ShortcutSlot slots[ZAPP_MAX_SHORTCUTS] = {0};
static UInt32 next_id = 1;
static EventHandlerRef carbon_handler = NULL;

static ShortcutSlot* find_slot_by_accelerator(const char* accelerator) {
    if (!accelerator) return NULL;
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (slots[i].active && strcmp(slots[i].accelerator, accelerator) == 0) {
            return &slots[i];
        }
    }
    return NULL;
}

static ShortcutSlot* find_slot_by_id(UInt32 id) {
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (slots[i].active && slots[i].id == id) return &slots[i];
    }
    return NULL;
}

static ShortcutSlot* claim_free_slot(void) {
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (!slots[i].active) return &slots[i];
    }
    return NULL;
}

// --- Carbon event handler ---

static OSStatus shortcut_event_handler(EventHandlerCallRef nextHandler,
                                        EventRef event, void* userData) {
    (void)nextHandler; (void)userData;
    EventHotKeyID hkID;
    OSStatus s = GetEventParameter(event, kEventParamDirectObject,
                                   typeEventHotKeyID, NULL,
                                   sizeof(hkID), NULL, &hkID);
    if (s != noErr) return s;
    ShortcutSlot* slot = find_slot_by_id(hkID.id);
    if (!slot) return noErr;

    // Build {"accelerator":"..."} payload, single-quoted-safe — the
    // JS template wraps the payload in single quotes.
    char escaped[256];
    {
        const char* src = slot->accelerator;
        size_t j = 0;
        for (size_t i = 0; src[i] && j + 2 < sizeof(escaped); i++) {
            char c = src[i];
            if (c == '\\' || c == '"' || c == '\'') escaped[j++] = '\\';
            escaped[j++] = c;
        }
        escaped[j] = '\0';
    }
    char payload[320];
    snprintf(payload, sizeof(payload), "{\"accelerator\":\"%s\"}", escaped);

    // Dispatch as a regular bridge event so both webviews and workers
    // receive it via Events.on. Same wire shape as deep-link OPEN_URL.
    char js[512];
    snprintf(js, sizeof(js),
        "(function(){var b=globalThis[Symbol.for('zapp.bridge')];"
        "if(b&&b._onEvent)b._onEvent('app:shortcut-triggered','%s');})();",
        payload);
    darwin_webview_eval_all(js);
    worker_broadcast_eval_js(js);

    return noErr;
}

static void ensure_carbon_handler_installed(void) {
    if (carbon_handler) return;
    EventTypeSpec spec = { kEventClassKeyboard, kEventHotKeyPressed };
    InstallApplicationEventHandler(&shortcut_event_handler, 1, &spec, NULL,
                                   &carbon_handler);
}

// --- Public API ---

bool darwin_shortcut_register(const char* accelerator) {
    if (!accelerator || !accelerator[0]) return false;
    @autoreleasepool {
        if (find_slot_by_accelerator(accelerator)) return false;  // already registered
        UInt32 mods = 0, key = 0;
        if (!parse_accelerator(accelerator, &mods, &key)) return false;

        ShortcutSlot* slot = claim_free_slot();
        if (!slot) return false;

        ensure_carbon_handler_installed();

        EventHotKeyID hkID = { .signature = 'zapp', .id = next_id++ };
        EventHotKeyRef ref = NULL;
        OSStatus status = RegisterEventHotKey(key, mods, hkID,
                                              GetApplicationEventTarget(),
                                              0, &ref);
        if (status != noErr || !ref) return false;

        slot->active = YES;
        slot->id = hkID.id;
        slot->ref = ref;
        strncpy(slot->accelerator, accelerator, sizeof(slot->accelerator) - 1);
        slot->accelerator[sizeof(slot->accelerator) - 1] = '\0';
        return true;
    }
}

bool darwin_shortcut_unregister(const char* accelerator) {
    ShortcutSlot* slot = find_slot_by_accelerator(accelerator);
    if (!slot) return false;
    if (slot->ref) UnregisterEventHotKey(slot->ref);
    slot->active = NO;
    slot->ref = NULL;
    slot->id = 0;
    slot->accelerator[0] = '\0';
    return true;
}

bool darwin_shortcut_is_registered(const char* accelerator) {
    return find_slot_by_accelerator(accelerator) != NULL;
}

void darwin_shortcut_unregister_all(void) {
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (slots[i].active && slots[i].ref) {
            UnregisterEventHotKey(slots[i].ref);
        }
        slots[i].active = NO;
        slots[i].ref = NULL;
        slots[i].id = 0;
        slots[i].accelerator[0] = '\0';
    }
}

File: native/platform/windows/shortcuts.c
// Windows global shortcuts — RegisterHotKey wrappers. Mirrors
// darwin/shortcuts.m: same accelerator grammar, same slot table, same
// app:shortcut-triggered bridge event on fire.

#define WIN32_LEAN_AND_MEAN
#include <windows.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "shortcuts.h"

extern void windows_webview_eval_all(const char* js);
extern void worker_broadcast_eval_js(char* js);

// --- Accelerator parsing ---

// Modifier token → MOD_* flag; 0 on unknown. "CmdOrCtrl" maps to Ctrl
// (one accelerator string working on both platforms is the point);
// explicit Cmd/Meta/Super pick the Windows key.
static UINT mod_from_token(const char* tok) {
    if (_stricmp(tok, "cmdorctrl") == 0 || _stricmp(tok, "commandorcontrol") == 0) return MOD_CONTROL;
    if (_stricmp(tok, "ctrl") == 0 || _stricmp(tok, "control") == 0) return MOD_CONTROL;
    if (_stricmp(tok, "cmd") == 0 || _stricmp(tok, "command") == 0 ||
        _stricmp(tok, "meta") == 0 || _stricmp(tok, "super") == 0) return MOD_WIN;
    if (_stricmp(tok, "alt") == 0 || _stricmp(tok, "option") == 0) return MOD_ALT;
    if (_stricmp(tok, "shift") == 0) return MOD_SHIFT;
    return 0;
}

// Key token → virtual-key code; 0 on unknown.
static UINT vk_from_token(const char* tok) {
    size_t len = strlen(tok);
    if (len == 0) return 0;
    if (len == 1) {
        char c = (char)toupper((unsigned char)tok[0]);
        if ((c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')) return (UINT)c;
        return 0;
    }
    if (_stricmp(tok, "space") == 0) return VK_SPACE;
    if (_stricmp(tok, "tab") == 0) return VK_TAB;
    if (_stricmp(tok, "return") == 0 || _stricmp(tok, "enter") == 0) return VK_RETURN;
    if (_stricmp(tok, "escape") == 0 || _stricmp(tok, "esc") == 0) return VK_ESCAPE;
    // darwin maps delete/backspace to the mac Delete key (= backspace);
    // forwarddelete is the del-below-insert key. Mirror that.
    if (_stricmp(tok, "delete") == 0 || _stricmp(tok, "backspace") == 0) return VK_BACK;
    if (_stricmp(tok, "forwarddelete") == 0) return VK_DELETE;
    if (_stricmp(tok, "left") == 0) return VK_LEFT;
    if (_stricmp(tok, "right") == 0) return VK_RIGHT;
    if (_stricmp(tok, "up") == 0) return VK_UP;
    if (_stricmp(tok, "down") == 0) return VK_DOWN;
    if (_stricmp(tok, "home") == 0) return VK_HOME;
    if (_stricmp(tok, "end") == 0) return VK_END;
    if (_stricmp(tok, "pageup") == 0) return VK_PRIOR;
    if (_stricmp(tok, "pagedown") == 0) return VK_NEXT;
    if ((tok[0] == 'f' || tok[0] == 'F') && len <= 3) {
        int n = atoi(tok + 1);
        if (n >= 1 && n <= 24) return VK_F1 + (UINT)(n - 1);
    }
    return 0;
}

// Parse "CmdOrCtrl+Shift+Space" → (modifiers, vk). False on any
// unknown component.
static bool parse_accelerator(const char* accelerator, UINT* out_mods, UINT* out_vk) {
    if (!accelerator || !accelerator[0]) return false;
    char buf[128];
    strncpy(buf, accelerator, sizeof(buf) - 1);
    buf[sizeof(buf) - 1] = '\0';

    UINT mods = 0;
    UINT vk = 0;
    char* tok = buf;
    while (tok) {
        char* sep = strchr(tok, '+');
        if (sep) *sep = '\0';
        // Trim surrounding whitespace.
        while (*tok == ' ' || *tok == '\t') tok++;
        char* end = tok + strlen(tok);
        while (end > tok && (end[-1] == ' ' || end[-1] == '\t')) *--end = '\0';

        if (sep) {
            UINT m = mod_from_token(tok);
            if (m == 0) return false;
            mods |= m;
            tok = sep + 1;
        } else {
            vk = vk_from_token(tok);
            tok = NULL;
        }
    }
    if (vk == 0) return false;
    *out_mods = mods;
    *out_vk = vk;
    return true;
}

// --- Registration table (mirrors darwin's ShortcutSlot) ---

#ifndef ZAPP_MAX_SHORTCUTS
#define ZAPP_MAX_SHORTCUTS 64
#endif

typedef struct {
    int active;
    int id; // RegisterHotKey id
    char accelerator[128];
} ShortcutSlot;

static ShortcutSlot slots[ZAPP_MAX_SHORTCUTS] = {0};
static int next_id = 1;

static ShortcutSlot* find_slot_by_accelerator(const char* accelerator) {
    if (!accelerator) return NULL;
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (slots[i].active && strcmp(slots[i].accelerator, accelerator) == 0) {
            return &slots[i];
        }
    }
    return NULL;
}

static ShortcutSlot* find_slot_by_id(int id) {
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (slots[i].active && slots[i].id == id) return &slots[i];
    }
    return NULL;
}

// --- Public API ---

bool windows_shortcut_register(const char* accelerator) {
    if (!accelerator || !accelerator[0]) return false;
    if (find_slot_by_accelerator(accelerator)) return false; // already registered
    UINT mods = 0, vk = 0;
    if (!parse_accelerator(accelerator, &mods, &vk)) return false;

    ShortcutSlot* slot = NULL;
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (!slots[i].active) { slot = &slots[i]; break; }
    }
    if (!slot) return false;

    int id = next_id++;
    // NULL hwnd → WM_HOTKEY lands on this thread's message queue;
    // platform.c's loop routes it to windows_shortcut_handle_wm_hotkey.
    // MOD_NOREPEAT: fire once per press, not on key-repeat (matches
    // Carbon's kEventHotKeyPressed semantics).
    if (!RegisterHotKey(NULL, id, mods | MOD_NOREPEAT, vk)) {
        return false; // usually: another app owns the combination
    }
    slot->active = 1;
    slot->id = id;
    strncpy(slot->accelerator, accelerator, sizeof(slot->accelerator) - 1);
    slot->accelerator[sizeof(slot->accelerator) - 1] = '\0';
    return true;
}

bool windows_shortcut_unregister(const char* accelerator) {
    ShortcutSlot* slot = find_slot_by_accelerator(accelerator);
    if (!slot) return false;
    UnregisterHotKey(NULL, slot->id);
    slot->active = 0;
    return true;
}

bool windows_shortcut_is_registered(const char* accelerator) {
    return find_slot_by_accelerator(accelerator) != NULL;
}

void windows_shortcut_unregister_all(void) {
    for (int i = 0; i < ZAPP_MAX_SHORTCUTS; i++) {
        if (slots[i].active) {
            UnregisterHotKey(NULL, slots[i].id);
            slots[i].active = 0;
        }
    }
}

// --- Fire dispatch ---

void windows_shortcut_handle_wm_hotkey(int hotkey_id) {
    ShortcutSlot* slot = find_slot_by_id(hotkey_id);
    if (!slot) return;

    // Same wire shape as darwin: app:shortcut-triggered with an
    // {"accelerator": "..."} payload, single-quote-escaped for the JS
    // string literal.
    char escaped[256];
    {
        const char* src = slot->accelerator;
        size_t j = 0;
        for (size_t i = 0; src[i] && j + 2 < sizeof(escaped); i++) {
            char c = src[i];
            if (c == '\\' || c == '"' || c == '\'') escaped[j++] = '\\';
            escaped[j++] = c;
        }
        escaped[j] = '\0';
    }
    char payload[320];
    snprintf(payload, sizeof(payload), "{\"accelerator\":\"%s\"}", escaped);

    char js[640];
    snprintf(js, sizeof(js),
        "(function(){var b=globalThis[Symbol.for('zapp.bridge')];"
        "if(b&&b._onEvent)b._onEvent('app:shortcut-triggered','%s');})();",
        payload);
    windows_webview_eval_all(js);
    worker_broadcast_eval_js(js);
}
